File: C++/Array_Easy_TwoSum_1.cpp
//Brute Force :

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target)
    {
        vector<int> solution_array(0);
        bool flag = false;

        for(int i = 0; i < nums.size(); i++)
        {
            for(int j = 0; j < nums.size(); j++)
            {
                if(nums[i] + nums[j] == target && i != j)
                {
                    solution_array.push_back(i);
                    solution_array.push_back(j);
                    flag = true;
                }
                if(flag == true)
                {
                    break;
                }
            }

            if(flag == true)
            {
                break;
            }
        }

        return solution_array;
    }
};

//Two Pass Hash Table : Optimal

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target)
    {
        unordered_map<int, int> the_map;
        vector<int> myVector;

        for(int i = 0; i < nums.size(); i++)
        {
            the_map[nums[i]] = i;
        }

        for(int j = 0; j < nums.size(); j++)
        {
            int complement = target - nums[j];
            int key = complement;
            auto it = the_map.find(key);

            if(it != the_map.end() && (*it).second != j)
            {
                myVector.push_back(j);
                myVector.push_back((*it).second);
                break;
            }
        }

        return myVector;
    }
};

File: Python/Array_Easy_TwoSum_1.py
# Basic Code TLE : 

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        array = []
        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                if nums[i] + nums[j] == target:
                    array = [i, j]
        return array
    
# Actual Two Sum Method : 

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        seen_dict = {}

        for i, number in enumerate(nums):
            complement = target - number

            if complement in seen_dict:
                return [seen_dict[complement], i]
            
            else:
                seen_dict[number] = i